### PR TITLE
Implement fuzz::wratio, process::{extract_one, extract_without_order} and add an initial set of unit tests

### PR DESCRIPTION
**Pull request by seanpianka**
With 15 stars, this is the most popular fuzzywuzzy port to Rust that I could find... but it's lacking process -- specifically extractOne -- which I'm looking to use in one of my Rust programs.

This PR adds an implementation for:
- `fuzz::wratio`
- `process::extract_one`
- `process::extract_without_order`

The `process` functions assume `choices` is a list, no "dict-like" (`HashMap`) choices are supported currently. Further, `process::extract_without_order` is missing key optimization steps to reduce calls to `full_process` and secondary query pre-processing.

I've also sprinkled in some function doc-strings, copying the original Python comments, along with my own in the code.

Lastly, I have added a subset of the original unit tests for, at the very least, basic code coverage. However, `fuzz::partial_token_sort_ratio` is failing one of its unit tests, and I believe a bug in the implementation is the root cause for failures in my `process::extract_one` unit tests.

### CODE REVIEW

**Comment by seanpianka** [comment]
Here are the test-case results as it stands today:
```
$ cargo test

running 16 tests
test fuzz::tests::test_empty_string_score_100 ... ok
test fuzz::tests::test_partial_ratio ... ok
test fuzz::tests::test_case_insensitive ... ok
test fuzz::tests::test_equal ... ok
test fuzz::tests::test_token_sort_ratio ... ok
test fuzz::tests::test_wratio_equal ... ok
test fuzz::tests::test_wratio_case_insensitive ... ok
test fuzz::tests::test_partial_token_set_ratio ... ok
test fuzz::tests::test_token_set_ratio ... ok
test fuzz::tests::test_wratio_partial_match ... ok
test fuzz::tests::test_partial_token_sort_ratio ... FAILED
test fuzz::tests::test_wratio_misordered_match ... FAILED
test process::tests::process::test_get_best_choice4 ... ok
test process::tests::process::test_get_best_choice1 ... ok
test process::tests::process::test_get_best_choice3 ... FAILED
test process::tests::process::test_get_best_choice2 ... ok

failures:

---- fuzz::tests::test_partial_token_sort_ratio stdout ----
thread 'fuzz::tests::test_partial_token_sort_ratio' panicked at 'assertion failed: `(left == right)`
  left: `67`,
 right: `100`', src/fuzz.rs:325:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- fuzz::tests::test_wratio_misordered_match stdout ----
thread 'fuzz::tests::test_wratio_misordered_match' panicked at 'assertion failed: `(left == right)`
  left: `95`,
 right: `50`', src/fuzz.rs:365:9

---- process::tests::process::test_get_best_choice3 stdout ----
thread 'process::tests::process::test_get_best_choice3' panicked at 'assertion failed: `(left == right)`
  left: `"braves vs mets"`,
 right: `"philladelphia phillies vs atlanta braves"`', src/process.rs:100:13


failures:
    fuzz::tests::test_partial_token_sort_ratio
    fuzz::tests::test_wratio_misordered_match
    process::tests::process::test_get_best_choice3

test result: FAILED. 13 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out
```

@logannc, I would appreciate your help debugging the issues with `fuzz::partial_token_sort_ratio`, as I believe it may be the root cause for `process::extract_one` test-case failures.

**Comment by logannc** [comment]
First off, thank you for the interest, the proactive PR, and generally being the kind of person who sees a need and fills it. :)

As I mentioned in our email, I'll try to take a look at this in the next few days. I have family in town which makes my free time a little unpredictable.

Also, this code is old and I'm a little ashamed of it, so you've inspired me to fix it up a bit and try to recall how to publish the crate to crates.io. :)

**Comment by logannc** [comment]
There is a bug in partial_ratio. I'm going to make a PR that fixes it with code that is almost verbatim the python code of fuzzywuzzy. After this is merged with - :ahem: - testing, I'll do a cleanup pass to make it more rust-y.

For the `process::extract_one` bug, it's still broken but for a more fun reason! Python and Rust have different semantics for which maximum value from an iterator is returned when there are multiple equal values!

> If multiple items are maximal, the function returns the first one encountered.
> [('new york mets vs chicago cubs', 26), ('chicago cubs vs chicago white sox', 36), ('philladelphia phillies vs atlanta braves', 86), ('braves vs mets', 86)]

https://docs.python.org/3/library/functions.html#max

> If several elements are equally maximum, the last element is returned.
> [("new york mets vs chicago cubs", 29),("chicago cubs vs chicago white sox", 31), ("philladelphia phillies vs atlanta braves", 86), ("braves vs mets", 86)]

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.max_by

That said they still have different scores on the first two values so not everything is perfect.

**Comment by logannc** [comment]
Okay, https://github.com/logannc/fuzzyrusty/pull/5 should fix one of your tests. For the extract test, you probably just need to reverse the iterator if you want to be 100% compatible with fuzzywuzzy. 

For the wratio test, I'm gunna let you take a look at that given the scoring differences. I wouldn't be *surprised* if the bug is in something I wrote (again, the shame), but what you added is sufficiently complex that it might be there.

**Comment by seanpianka** [comment]
I appreciate you quickly responding after only having last touched this code four years ago! I'm glad to know you're still here 👍

This slight difference in behavior between the two approaches of finding maximum values is quite the tricky bug, good catch spotting that. In my latest commits, I've rebased off of your changes from #5 and added the reversal of the iterator, along with your description of the issue (as I think it's optimal that a fuzzywuzzy port should behavior as similar as possible to the original).

However, I'm still seeing failures in the `fuzz::...` tests:
```rust

running 16 tests
test fuzz::tests::test_partial_ratio ... ok
test fuzz::tests::test_case_insensitive ... ok
test fuzz::tests::test_empty_string_score_100 ... ok
test fuzz::tests::test_token_sort_ratio ... ok
test fuzz::tests::test_equal ... ok
test fuzz::tests::test_wratio_case_insensitive ... ok
test fuzz::tests::test_wratio_equal ... ok
test fuzz::tests::test_partial_token_set_ratio ... ok
test fuzz::tests::test_token_set_ratio ... ok
test fuzz::tests::test_partial_token_sort_ratio ... FAILED
test fuzz::tests::test_wratio_misordered_match ... FAILED
test fuzz::tests::test_wratio_partial_match ... ok
test process::tests::process::test_get_best_choice1 ... ok
test process::tests::process::test_get_best_choice4 ... ok
test process::tests::process::test_get_best_choice2 ... ok
test process::tests::process::test_get_best_choice3 ... ok

failures:

---- fuzz::tests::test_partial_token_sort_ratio stdout ----
thread 'fuzz::tests::test_partial_token_sort_ratio' panicked at 'assertion failed: `(left == right)`
  left: `50`,
 right: `100`', src/fuzz.rs:371:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- fuzz::tests::test_wratio_misordered_match stdout ----
thread 'fuzz::tests::test_wratio_misordered_match' panicked at 'assertion failed: `(left == right)`
  left: `95`,
 right: `50`', src/fuzz.rs:414:9


failures:
    fuzz::tests::test_partial_token_sort_ratio
    fuzz::tests::test_wratio_misordered_match

test result: FAILED. 14 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out
```

If you check out [the original WRatio implementation](https://github.com/seatgeek/fuzzywuzzy/blob/master/fuzzywuzzy/fuzz.py#L224), you'll see that `fuzz::wratio` is almost "word-for-word" to the original code (intentionally, hopefully it reads similarly). I'll review it again to make sure it's ported correctly, but I have an itch that it may be root-caused by the cause of partial_token_sort_ratio test failures because `fuzz::wratio` invokes `fuzz::partial_token_sort_ratio`.

Thanks again for the timely response!

**Comment by logannc** [comment]
Oh, right. I think you copied the wrong expected value for the partial token sort ratio test. I couldn't fix that in my PR because the value didn't exist! Oops!

File: src/fuzz.rs
Comment (logannc) [inline src/fuzz.rs]:
```suggestion
            50
```

Comment (seanpianka) [reply src/fuzz.rs]: Great catch, [source](https://github.com/seatgeek/fuzzywuzzy/blob/master/test_fuzzywuzzy.py#L135). Adding this suggestion.

**Comment by logannc** [comment]
I'll take a deeper look tomorrow at where in wratio things diverge from the reference impl. As I said, it may still totally be my fault. This is not my highest quality code. ;)

File: src/fuzz.rs
Comment (logannc) [inline src/fuzz.rs]:
```suggestion
        assert_eq!(fuzz::wratio(f.s4, f.s5, true, true), 95);
```

found the bug ;)

File: src/utils.rs
Comment (logannc) [inline src/utils.rs]:
```suggestion
        result = result.chars().filter(char::is_ascii).collect();
```

File: src/utils.rs
Comment (logannc) [inline src/utils.rs:21]: I think this can probably be inlined where it is called. I do not anticipate additions to this function that make it a worthwhile abstraction.

File: src/process.rs
Comment (logannc) [inline src/process.rs]: lets not print anything. we're a library. if we want to add some kind of logging, I think we'd want it to be opt in.

File: src/process.rs
Comment (logannc) [inline src/process.rs:28]: are these TODOs still relevant? if so, please create an issue and update the comment with a reference to them.

File: src/process.rs
Comment (logannc) [inline src/process.rs]: my rust is rusty so I'm not prepared to make a declaration on this without experiments. however, have you considered not using trait objects? I tested it and it works without modification to the call sites. I need to run some tests to see if its worth it.

We can also just leave this as is, right now. The crate is apparently published by someone that isn't me and I need to figure out how to get in touch with them, so even if changing this to a non-trait object later is a breaking change, we can probably do that before the next crate push.

File: src/process.rs
Comment (seanpianka) [reply src/process.rs]: I think it's a useful warning to have, so I'll comment this out and mark it as // TODO: to make sure warnings and/or verbose loggings are available options to be enabled when using the library. This addition can be done when the input parameters to `process::*` are moved to option structs to provide default parameters for `processor`, `scorer`, and `score_cutoff`.

File: src/process.rs
Comment (seanpianka) [reply src/process.rs:28]: Yes, they're issues that are not coupled with GitHub :) Are these two points from the original implementation optimizations (prevent duplicate work) or bugs (re-running processor function multiple times may produce unwanted transformed input data)?

File: src/process.rs
Comment (seanpianka) [reply src/process.rs]: I believe the main benefit of using trait objects is that arguments to `processor` can be closures which capture their environment (i.e. receive some sort of reference to `self`).

e.g.

Capturing environment with a closure:
```rust
fn main() {
    let x = 4;
    let equal_to_x = |z| z == x;
    let y = 4;
    assert!(equal_to_x(y));
}
```

With a function:
```rust
fn main() {
    let x = 4;
    fn equal_to_x(z: i32) -> bool {
        z == x
    }
    let y = 4;
    assert!(equal_to_x(y));
}

$ cargo run
   Compiling equal-to-x v0.1.0 (file:///projects/equal-to-x)
error[E0434]: can't capture dynamic environment in a fn item
 --> src/main.rs:5:14
  |
5 |         z == x
  |              ^
  |
  = help: use the `|| { ... }` closure form instead

error: aborting due to previous error

For more information about this error, try `rustc --explain E0434`.
error: could not compile `equal-to-x`.

To learn more, run the command again with --verbose.
```

Using the trait object [`Fn`](https://doc.rust-lang.org/std/ops/trait.Fn.html) allows for arguments to be closures which capture their environment, while [primitive function pointers](https://doc.rust-lang.org/std/primitive.fn.html) do not support this type of argument.

File: src/utils.rs
Comment (seanpianka) [reply src/utils.rs:21]: I think there is value _in case of_ future, unanticipated additions, along with maintaining a similarity in design to the original Python implementation, which will help in remaining accessible to new contributors.

File: src/process.rs
Comment (logannc) [reply src/process.rs]: I think this is untrue. With your branch checked out, changing the signature for processor to `U ... where U: Fn(&str, bool) -> String` in `extract_one` and `extract_without_order` allows me to do the following in `unwrap_extract_one_choice`:

```rust
let b = true;
extract_one(
    query,
    get_baseball_strings().iter(),
    |a, _| utils::full_process(a, b),
    &fuzz::wratio,
    0)
... etc
```

It allows: a) function references b) functions by value (i.e., `utils::full_process` without `&`) and c) closures with or without captures.

If you don't want to change it, feel free to resolve this. But I will look into making the change before publishing the crate.

**Review by logannc — APPROVED** [review]

**Comment by logannc** [comment]
I may follow up on some things after merge, but don't want to hold this up. Let me know if you do not intend to make the function type change and we can merge.

Thank you for this!

File: src/process.rs
Comment (seanpianka) [reply src/process.rs]: Okay, I've replaced the usages of trait objects with trait bounds. There is no compelling reason to use trait objects here, it would seem.

File: src/process.rs
Comment (seanpianka) [reply src/process.rs:28]: Added #6 and mentioned it above in the comments

**Comment by seanpianka** [comment]
> I may follow up on some things after merge, but don't want to hold this up. Let me know if you do not intend to make the function type change and we can merge.
> 
> Thank you for this!

With pleasure!

Please merge when you are ready. Even though you approved the PR,  I do not have write access to this repository, so you will have to merge.

**Pull request merged**
